pystiche/papers/sanakoyeu_et_al_2018/core.py

```diff
@@ -0,0 +1,302 @@
+import itertools
+from typing import Callable, Optional, Union
+
+import torch
+from torch import nn
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+from torch.optim.optimizer import Optimizer
+from torch.utils.data import DataLoader
+
+from pystiche.loss.perceptual import PerceptualLoss
+from pystiche.ops.container import MultiLayerEncodingOperator
+
+from .data import (
+    sanakoyeu_et_al_2018_dataset,
+    sanakoyeu_et_al_2018_image_loader,
+    sanakoyeu_et_al_2018_images,
+)
+from .loss import (
+    MultiLayerDicriminatorEncodingOperator,
+    SanakoyeuEtAl2018DiscriminatorLoss,
+    sanakoyeu_et_al_2018_discriminator_operator,
+    sanakoyeu_et_al_2018_transformer_loss,
+)
+from .modules import SanakoyeuEtAl2018Discriminator, SanakoyeuEtAl2018Transformer
+from .utils import (
+    ExponentialMovingAverage,
+    sanakoyeu_et_al_2018_lr_scheduler,
+    sanakoyeu_et_al_2018_optimizer,
+)
+
+__all__ = [
+    "sanakoyeu_et_al_2018_training",
+    "sanakoyeu_et_al_2018_stylization",
+    "SanakoyeuEtAl2018Transformer",
+    "SanakoyeuEtAl2018Discriminator",
+    "sanakoyeu_et_al_2018_transformer_loss",
+    "SanakoyeuEtAl2018DiscriminatorLoss",
+    "sanakoyeu_et_al_2018_discriminator_operator",
+    "sanakoyeu_et_al_2018_images",
+    "sanakoyeu_et_al_2018_image_loader",
+    "sanakoyeu_et_al_2018_dataset",
+]
+
+
+def gan_optim_loop(
```

Comment:
We should namespace this with authors name etc. since these shenanigans with also using the `content_image` directly is most likely not a general case.